### PR TITLE
feat(editor): add Bear backup import and markdown zip folder hierarchy

### PR DESCRIPTION
**Pull request by karl-kaefer**
## Summary

- Add Bear `.bear2bk` backup importer (TextBundle-based zip format)
- Enhance markdown zip import to preserve folder structure from zip paths
- Add colored highlight (`<mark data-color="...">`) support to HTML adapter

### Bear Import Details

Bear backups are zip archives of TextBundle directories. The importer:
- Parses Bear-specific markdown (highlights `==text==`, callouts `> [!NOTE]`, inline tags `#tag`)
- Extracts creation/modification dates from `info.json` metadata
- Filters out trashed notes
- Converts Bear tags to AFFiNE tags (consolidated by root segment)
- Builds folder hierarchy from nested tag paths (e.g., `#work/projects/alpha`)
- Uses JSZip for lazy decompression to handle large backups without OOM

### Markdown Zip Folder Hierarchy

`importMarkdownZip` now returns `{ docIds, folderHierarchy }` instead of just `docIds[]`, enabling the UI to recreate the zip's directory structure as AFFiNE folders.

## Related Issues

- Implements the TextBundle-based import approach suggested in #14115 / Discussion #14142
- Addresses folder structure preservation requested in #10003
- Partially addresses frontmatter metadata import from #11286

## Test Plan

- [ ] Import a Bear `.bear2bk` backup file via the import dialog
- [ ] Verify tags are created and assigned to documents
- [ ] Verify folder hierarchy matches Bear's nested tag structure
- [ ] Verify creation/modification dates are preserved
- [ ] Verify highlighted text and callouts render correctly
- [ ] Verify images and attachments are imported
- [ ] Import a markdown zip with nested folders, verify folder structure is recreated
- [ ] Verify trashed Bear notes are excluded

## Summary by CodeRabbit

* **New Features**
  * Bear (.bear2bk) backup import: bulk import notes, convert/dedupe tags, create nested folders, and return imported doc IDs plus folder hierarchy; UI import option and progress integrated.
  * Markdown ZIP import now returns an optional folder hierarchy alongside created doc IDs.

* **Bug Fixes / Improvements**
  * Highlighting: mark elements validate color names, default safely, and apply consistent background styling.

* **Chores**
  * Added runtime dependency for ZIP handling.

* **Documentation**
  * Added localization strings and i18n accessors for Bear import UI.

### CODE REVIEW

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+      lines[i] = lines[i].replace(
+        /==(\S(?:[^=]|=[^=])*?)==/g,
+        (_match, content: string) => {
+          const firstChar = String.fromCodePoint(content.codePointAt(0)!);
+          const color = HIGHLIGHT_COLOR_MAP[firstChar];
+          if (color) {
+            const text = content.slice(firstChar.length);
+            return `<mark data-color="${color}">${text}</mark>`;
+          }
+          return `<mark>${content}</mark>`;
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts]: Fixed in fa008195b. Added an `escapeHtml` utility and applied it to both the colored and plain `<mark>` branches in `convertHighlights`. This prevents markup injection from content like `==<script>alert(1)</script>==`.

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+    filteredLines.length > 0 &&
+    filteredLines[filteredLines.length - 1].trim() === ''
+  ) {
+    filteredLines.pop();
+  }
+
+  return {
+    tags: deduplicateTags(tags),
+    content: filteredLines.join('\n'),
+  };
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts:107]: This is intentional. Bear places tags at the end of notes as a footer convention. Scanning the entire document body for `#tags` would produce false positives from markdown headings (`# Title`), code comments, anchors, and other `#` prefixed content.

The reverse-scan approach matches how Bear actually stores tags: as dedicated tag-only lines at the bottom of the note. I verified this against a real Bear backup with 514 notes and it correctly extracted all tags.

If there are Bear notes that use mid-document inline tags (which would be unusual), those tags are also stored in the `info.json` metadata and could be read from there in a future enhancement.

File: blocksuite/affine/widgets/linked-doc/src/transformers/markdown.ts [hunk 3]
```diff
@@ -448,10 +460,89 @@ async function importMarkdownZip({
...
+        e.fullPath.startsWith(folderParts[0] + '/')
+      );
+      if (allSameRoot) {
+        folderParts = folderParts.slice(1);
+      }
+    }
+
+    if (folderParts.length === 0) {
+      // Root-level file, no folder needed
+      continue;
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/markdown.ts:597]: This is intentional. The `parts.length > 2` threshold exists because most zip tools wrap all files in a single root directory (e.g. `export/file1.md`, `export/file2.md`). With `parts.length >= 2`, every such zip would create a spurious top-level folder named after the zip archive itself.

The common-root stripping already handles removing that wrapper directory. After stripping, `root/file.md` correctly becomes a root-level file with no folder, which is the expected behavior. Only zips with real nested structure (`root/subfolder/file.md`) produce folders.

This was tested with zip archives of various layouts and the current behavior matches user expectations for markdown zip imports.

File: packages/frontend/core/src/desktop/dialogs/import/index.tsx [hunk 4]
```diff
@@ -445,6 +492,124 @@ const importConfigs: Record<ImportType, ImportConfig> = {
       };
     },
   },
+  bear: {
+    fileOptions: { acceptType: 'Zip', multiple: false },
+    importFunction: async (
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply packages/frontend/core/src/desktop/dialogs/import/index.tsx:545]: This is a known limitation of the Windows `showOpenFilePicker` API. I originally implemented a dedicated `'Bear'` acceptType with `['.zip', '.bear2bk']`, but Windows silently ignores custom/unknown extensions in the file picker filter, making `.bear2bk` files invisible even when the filter is active.

Using `'Zip'` as the acceptType works in practice because:
1. Bear `.bear2bk` files are just renamed `.zip` files
2. Users can switch the file dialog filter to "All files" to select `.bear2bk` directly
3. This was successfully tested with a real 194MB Bear backup (514 notes, 327 assets)

A better solution would require changes to the `openFilesWith` infrastructure in `filesys.ts` to handle platform-specific quirks, which is out of scope for this PR. The fallback `<input>` path on browsers without `showOpenFilePicker` does not have this limitation.

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts]: The new comment about `convertGfmCallouts` not skipping code blocks is also fixed in f66c2b3e8. The function now processes line-by-line with code fence tracking, matching the approach already used in `convertHighlights`.

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+    /^(>\s*)\[!(\w+)\]/gm,
+    (_match, prefix: string, type: string) => {
+      const emoji = GFM_CALLOUT_MAP[type.toUpperCase()];
+      if (emoji) {
+        return `${prefix}[!${emoji}]`;
+      }
+      return _match;
+    }
+  );
+}
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts:261]: Fixed in f66c2b3e8. `convertGfmCallouts` now processes line-by-line with code fence tracking, matching the approach used in `convertHighlights`. Literal fenced examples containing `> [!NOTE]` are left untouched.

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+        const ts = Date.parse(String(bearMeta.creationDate));
+        if (!isNaN(ts)) metaPatch.createDate = ts;
+      }
+      if (bearMeta?.modificationDate) {
+        const ts = Date.parse(String(bearMeta.modificationDate));
+        if (!isNaN(ts)) metaPatch.updatedDate = ts;
+      }
+      if (Object.keys(metaPatch).length) {
+        collection.meta.setDocMeta(doc.id, metaPatch);
+      }
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts]: Fixed in f66c2b3e8. The per-bundle import loop is now wrapped in try/catch. A single failing note logs a warning and continues to the next bundle instead of aborting the entire import. Successfully imported docs remain in the workspace.

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+
+  // Process bundles sequentially to limit memory.
+  // Each bundle is wrapped in try/catch so one bad note does not abort the
+  // entire import after earlier notes have already been written.
+  for (const { entry, bearMeta } of validBundles) {
+    try {
+      // Read markdown (decompress on demand)
+      const rawMarkdown = await zip.file(entry.markdownPath!)!.async('string');
+      if (!rawMarkdown.trim()) continue;
+
```
[the remark this answers is not in the document]
Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts:423]: I'm not going to fix this one. Here's why:
A Bear note with a completely empty body (`rawMarkdown.trim() === ''`) has no content, no title, and no tags to import. Passing empty markdown to the adapter would create a blank untitled document in the workspace: that's not useful, it's clutter. 
Bear itself allows empty notes to exist, but they carry no meaningful data worth importing. This is the same reasoning as before: the skip is intentional. In my real-world test of 514 Bear notes, zero were empty - this is an edge case that would only produce garbage docs if "fixed."

**Comment by karl-kaefer** [comment]
@darkskygit Could you please re-run the failed CI jobs? Both failures appear unrelated to this PR. The share-page E2E test   ([share-page-1.spec.ts:210](https://github.com/toeverything/AFFiNE/blob/canary/tests/affine-cloud/e2e/share-page-1.spec.ts#L210)) also fails on canary (run [24001775690](https://github.com/toeverything/AFFiNE/actions/runs/24001775690)), and the server config test is a Redis teardown issue. 

cc @ibex088 since your #14756 touched the failing test.

**Comment by karl-kaefer** [comment]
Sorry for pinging you again @darkskygit - could you please re-trigger? forgot to push the changes. Sorry!

**Comment by karl-kaefer** [comment]
@darkskygit - this PR is ready for review. CI is green after rebasing onto latest canary (the failure before was a random flaky CI env glitch). 
The PR adds a Bear notes import via .bear2bk (Bear backup zipped MD file) with tag mapping, folder hierarchy and highlight/callout conversion.
Happy to address any feedback.

File: blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+/** Deduplicate tags case-insensitively, normalizing to lowercase. */
+function deduplicateTags(tags: string[]): string[] {
+  const seen = new Set<string>();
+  const result: string[] = [];
+  for (const tag of tags) {
+    const normalized = tag.toLowerCase();
+    if (!seen.has(normalized)) {
+      seen.add(normalized);
+      result.push(normalized);
```
Comment (darkskygit) [inline blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts]: It might be better to keep the original capitalization of the tag?

Comment (karl-kaefer) [reply blocksuite/affine/widgets/linked-doc/src/transformers/bear.ts]: Good catch, fixed in c46e88a73. Tag capitalization is now preserved (first-seen casing) while still deduping case-insensitively, and the same fix is applied to the root-tag consolidation in the import dialog so new AFFiNE tags keep the user's original capitalization.

**Review by darkskygit — APPROVED** [review]
